hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/datanode/fsdataset/impl/PmemVolumeManager.java

```diff
@@ -283,52 +283,32 @@ void cleanup() {
    * Restore cache from the cached files in the configured pmem volumes.
    */
   public Map<ExtendedBlockId, MappableBlock> restoreCache(
-      String bpid, boolean isNative) throws IOException {
+      String bpid, MappableBlockLoader cacheLoader) throws IOException {
     final Map<ExtendedBlockId, MappableBlock> keyToMappableBlock
         = new ConcurrentHashMap<>();
     for (byte volumeIndex = 0; volumeIndex < pmemVolumes.size();
          volumeIndex++) {
       long maxBytes = usedBytesCounts.get(volumeIndex).getMaxBytes();
       long usedBytes = 0;
       File cacheDir = new File(pmemVolumes.get(volumeIndex), bpid);
-      if (cacheDir.exists()) {
-        Collection<File> cachedFileList = FileUtils.listFiles(cacheDir,
-            TrueFileFilter.INSTANCE, TrueFileFilter.INSTANCE);
-        // Scan the cached files in pmem volumes for cache restore.
-        for (File cachedFile : cachedFileList) {
-          usedBytes += cachedFile.length();
-          String[] bid = cachedFile.getName().split("-");
-          String blockId = bid[bid.length - 1];
-          String blockPoolId = cachedFile.getName()
-              .replace("-" + blockId, "");
-          if (!blockPoolId.equals(bpid)) {
-            throw new IOException("Illegal cache file name, " +
-                "it should be named by BlockPoolId-BlockId");
-          }
-          ExtendedBlockId key = new ExtendedBlockId(
-              Long.parseLong(blockId), bpid);
-          blockKeyToVolume.put(key, volumeIndex);
-          if (!isNative) {
-            keyToMappableBlock.put(key, new PmemMappedBlock(
-                cachedFile.length(), key));
-          } else {
-            NativeIO.POSIX.PmemMappedRegion region =
-                NativeIO.POSIX.Pmem.mapBlock(cachedFile.getAbsolutePath(),
-                    cachedFile.length(), true);
-            if (region == null) {
-              throw new IOException("Failed to restore the block "
-                  + cachedFile.getName() + " in persistent storage.");
-            }
-            keyToMappableBlock.put(key, new NativePmemMappedBlock(
-                    region.getAddress(), region.getLength(), key));
-          }
-        }
-        // Update maxBytes and cache capacity according to cache space
-        // used by restored cached files.
-        usedBytesCounts.get(volumeIndex).setMaxBytes(maxBytes + usedBytes);
-        cacheCapacity += usedBytes;
-        usedBytesCounts.get(volumeIndex).reserve(usedBytes);
+      Collection<File> cachedFileList = FileUtils.listFiles(cacheDir,
+          TrueFileFilter.INSTANCE, TrueFileFilter.INSTANCE);
+      // Scan the cached files in pmem volumes for cache restore.
+      for (File cachedFile : cachedFileList) {
+        usedBytes += cachedFile.length();
```

Comment:
move this statement to the place after getMappableBlock(), where the cache is recovered. Thus, if there is exception during the restore, the usedBytes will not count the file size.